src/tracer.cc

```diff
@@ -617,6 +658,10 @@ TraceManager::TraceActivity(
       reinterpret_cast<std::shared_ptr<TraceManager::Trace>*>(userp)->get();
 
   std::lock_guard<std::mutex> lk(ts->mtx_);
+  if (ts->spawned_traces_tracker_.find(id) ==
+      ts->spawned_traces_tracker_.end()) {
+    ts->spawned_traces_tracker_.emplace(id);
+  }
```

Comment:
How about adding new `TraceSpawn` callback for the original trace object that the user may add to the trace, which will be invoked during `TRITONSERVER_InferenceTraceSpawnChildTrace` if set. The user may use it to trace spawned traces rather than implementing the logic in `TraceActivity` callback.

However, if we were providing the new callback, I would expect it to be where the child trace to be created. In other words, instead of Triton internally creates the child trace, the user will need to do it explicitly in the callback:
```
TRITONSERVER_InferenceTraceSpawnFn_t(parent_trace, *child_trace) {
    TRITONSERVER_InferenceTraceTensorNew(child_trace, ...);
}
```
This is extra effort for the developer but it gives flexibility on how child trace is reported, i.e. the trace level / activity callback can be completely different from parent trace.

Reply:
Should we consider it as a follow up? It seems like a nice addition, but I feel like it needs to be discussed in more details.